### ISSUE TITLE
mountfsd fails compilation if zlib or xz are not available (hidden dependency on lib_import_common)

### ISSUE THREAD

**Issue by myrkr** [systemd/systemd#32572]
### systemd version the issue has been seen with

256-rc1

### Used distribution

None

### CPU architectures issue was seen on

x86_64

### Component

importd, mountfsd

### Expected behaviour you didn't see

systemd can be compiled without importd by disabling zlib and xz.

### Unexpected behaviour you saw

mountfsd has a hidden dependency on libimport-common.a. This dependency is not guarded by ENABLED_IMPORTD causing the following build error:

```
[112/1463] Compiling C object src/import/libimport-common.a.p/import-compress.c.o
FAILED: src/import/libimport-common.a.p/import-compress.c.o
ccache cc -Isrc/import/libimport-common.a.p -Isrc/import -I../src/import -Isrc/basic -I../src/basic -Isrc/fundamental -I../src/fundamental -Isrc/systemd -I../src/systemd -I. -I.. -I../src/libsystemd/sd-bus -I../src/libsystemd/sd-device -I../src/libsystemd/sd-event -I../src/libsystemd/sd-hwdb -I../src/libsystemd/sd-id128 -I../src/libsystemd/sd-journal -I../src/libsystemd/sd-netlink -I../src/libsystemd/sd-network -I../src/libsystemd/sd-resolve -Isrc/shared -I../src/shared -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -std=gnu11 -O0 -g -Wno-missing-field-initializers -Wno-unused-parameter -Wno-nonnull-compare -Warray-bounds -Warray-bounds=2 -Wdate-time -Wendif-labels -Werror=format=2 -Werror=format-signedness -Werror=implicit-function-declaration -Werror=implicit-int -Werror=incompatible-pointer-types -Werror=int-conversion -Werror=missing-declarations -Werror=missing-prototypes -Werror=overflow -Werror=override-init -Werror=return-type -Werror=shift-count-overflow -Werror=shift-overflow=2 -Werror=undef -Wfloat-equal -Wimplicit-fallthrough=5 -Winit-self -Wlogical-op -Wmissing-include-dirs -Wmissing-noreturn -Wnested-externs -Wold-style-definition -Wpointer-arith -Wredundant-decls -Wshadow -Wstrict-aliasing=2 -Wstrict-prototypes -Wsuggest-attribute=noreturn -Wunused-function -Wwrite-strings -Wzero-length-bounds -fdiagnostics-show-option -fno-common -fstack-protector -fstack-protector-strong --param=ssp-buffer-size=4 -Wno-unused-result -Werror=shadow -fPIC -fno-strict-aliasing -fvisibility=hidden -fno-omit-frame-pointer -include config.h -MD -MQ src/import/libimport-common.a.p/import-compress.c.o -MF src/import/libimport-common.a.p/import-compress.c.o.d -o src/import/libimport-common.a.p/import-compress.c.o -c ../src/import/import-compress.c
In file included from ../src/import/import-compress.c:3:
../src/import/import-compress.h:7:10: fatal error: lzma.h: No such file or directory
    7 | #include <lzma.h>
      |          ^~~~~~~~
compilation terminated.
[114/1463] Compiling C object src/import/libimport-common.a.p/qcow2-util.c.o
FAILED: src/import/libimport-common.a.p/qcow2-util.c.o
ccache cc -Isrc/import/libimport-common.a.p -Isrc/import -I../src/import -Isrc/basic -I../src/basic -Isrc/fundamental -I../src/fundamental -Isrc/systemd -I../src/systemd -I. -I.. -I../src/libsystemd/sd-bus -I../src/libsystemd/sd-device -I../src/libsystemd/sd-event -I../src/libsystemd/sd-hwdb -I../src/libsystemd/sd-id128 -I../src/libsystemd/sd-journal -I../src/libsystemd/sd-netlink -I../src/libsystemd/sd-network -I../src/libsystemd/sd-resolve -Isrc/shared -I../src/shared -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -std=gnu11 -O0 -g -Wno-missing-field-initializers -Wno-unused-parameter -Wno-nonnull-compare -Warray-bounds -Warray-bounds=2 -Wdate-time -Wendif-labels -Werror=format=2 -Werror=format-signedness -Werror=implicit-function-declaration -Werror=implicit-int -Werror=incompatible-pointer-types -Werror=int-conversion -Werror=missing-declarations -Werror=missing-prototypes -Werror=overflow -Werror=override-init -Werror=return-type -Werror=shift-count-overflow -Werror=shift-overflow=2 -Werror=undef -Wfloat-equal -Wimplicit-fallthrough=5 -Winit-self -Wlogical-op -Wmissing-include-dirs -Wmissing-noreturn -Wnested-externs -Wold-style-definition -Wpointer-arith -Wredundant-decls -Wshadow -Wstrict-aliasing=2 -Wstrict-prototypes -Wsuggest-attribute=noreturn -Wunused-function -Wwrite-strings -Wzero-length-bounds -fdiagnostics-show-option -fno-common -fstack-protector -fstack-protector-strong --param=ssp-buffer-size=4 -Wno-unused-result -Werror=shadow -fPIC -fno-strict-aliasing -fvisibility=hidden -fno-omit-frame-pointer -include config.h -MD -MQ src/import/libimport-common.a.p/qcow2-util.c.o -MF src/import/libimport-common.a.p/qcow2-util.c.o.d -o src/import/libimport-common.a.p/qcow2-util.c.o -c ../src/import/qcow2-util.c
../src/import/qcow2-util.c:3:10: fatal error: zlib.h: No such file or directory
    3 | #include <zlib.h>
      |          ^~~~~~~~
compilation terminated.
[145/1463] Linking target src/shared/libsystemd-shared-256.so
```

The problem is likely caused by `src/importd` providing a variable `common_libs`:

```
common_libs = [
        lib_import_common,
        libshared,
]
```

and `src/mountfsd` using that lib:

```
executables += [
        libexec_template + {
                'name' : 'systemd-mountfsd',
                'conditions' : ['ENABLE_MOUNTFSD'],
                'sources' : systemd_mountfsd_sources,
        },
        libexec_template + {
                'name' : 'systemd-mountwork',
                'conditions' : ['ENABLE_MOUNTFSD'],
                'sources' : systemd_mountwork_sources,
                'link_with' : common_libs,
                'dependencies' : common_deps,
        },
]
```

### Steps to reproduce the problem

Our buildenv does not have the development packages for zlib and xz installed. We build systemd with the following options:

```
    --buildtype=plain \
    --prefix=/$(PC_PREFIX) \
    -D libdir=/lib \
    -D dbuspolicydir=/etc/dbus-1/system.d \
    -D dbussessionservicedir=/usr/share/dbus-1/services \
    -D dbussystemservicedir=/usr/share/dbus-1/system-services \
    -D bashcompletiondir=/etc/bash_completion.d \
    -D time-epoch=946681200 \
    -D clock-valid-range-usec-max=$(VALID_CLOCK_PERIOD) \
    -D acl=false \
    -D apparmor=false \
    -D audit=false \
    -D backlight=false \
    -D binfmt=false \
    -D blkid=true \
    -D bzip2=false \
    -D certificate-root=/etc/ssl \
    -D coredump=true \
    -D dbus=false \
    -D default-dnssec=false \
    -D efi=false \
    -D elfutils=false \
    -D environment-d=false \
    -D fallback-hostname=localhost \
    -D firstboot=false \
    -D default-dnssec=no \
    -D dns-servers= \
    -D gcrypt=false \
    -D glib=false \
    -D gnutls=false \
    -D hibernate=false \
    -D hostnamed=false \
    -D html=false \
    -D hwdb=true \
    -D idn=false \
    -D ima=false \
    -D importd=false \
    -D kmod=true \
    -D ldconfig=false \
    -D libcryptsetup=false \
    -D libcurl=false \
    -D libidn=false \
    -D libiptc=false \
    -D localed=false \
    -D logind=true \
    -D lz4=false \
    -D man=false \
    -D microhttpd=false \
    -D nss-myhostname=true \
    -D networkd=true \
    -D nss-systemd=false \
    -D ntp-servers= \
    -D polkit=false \
    -D qrencode=false \
    -D quotacheck=false \
    -D randomseed=false \
    -D remote=false \
    -D resolve=false \
    -D rfkill=false \
    -D selinux=false \
    -D smack=false \
    -D sysusers=false \
    -D sysvinit-path= \
    -D sysvrcnd-path= \
    -D timedated=false \
    -D timesyncd=true \
    -D tmpfiles=false \
    -D tpm=false \
    -D utmp=false \
    -D vconsole=true \
    -D xkbcommon=false \
    -D xz=false \
    -D zlib=false \
    -D split-usr=true \
    -D openssl=false \
    -D tpm2=false \
    -D split-bin=true \

```
    
Here is the output of meson regarding the build options:
    
```
systemd 256~rc1

    build mode                               : developer
    split bin-sbin                           : True
    prefix directory                         : /
    sysconf directory                        : /etc
    include directory                        : /include
    lib directory                            : /lib
    SysV init scripts                        :
    SysV rc?.d directories                   :
    PAM modules directory                    : /lib/security
    PAM configuration directory              : /lib/pam.d
    ssh server configuration directory       : /etc/ssh/sshd_config.d
    ssh server privilege separation directory: /run/sshd
    ssh client configuration directory       : /etc/ssh/ssh_config.d
    libcryptsetup plugins directory          : /lib/cryptsetup
    RPM macros directory                     : /lib/rpm/macros.d
    modprobe.d directory                     : /lib/modprobe.d
    D-Bus policy directory                   : /etc/dbus-1/system.d
    D-Bus session directory                  : /usr/share/dbus-1/services
    D-Bus system directory                   : /usr/share/dbus-1/system-services
    D-Bus interfaces directory               : /share/dbus-1/interfaces
    bash completions directory               : /etc/bash_completion.d
    zsh completions directory                : /share/zsh/site-functions
    private shared lib version tag           : 256
    extra start script                       :
    debug shell                              : /bin/sh @ /dev/tty9
    system UIDs                              : <=999 (alloc >=1)
    system GIDs                              : <=999 (alloc >=1)
    dynamic UIDs                             : 61184…65519
    container UID bases                      : 524288…1878982656
    static UID/GID allocations               : tty-gid:5
    /dev/kvm access mode                     : 0666
    render group access mode                 : 0666
    certificate root directory               : /etc/ssl
    support URL                              : https://lists.freedesktop.org/mailman/listinfo/systemd-devel
    nobody user name                         : nobody
    nobody group name                        : nobody
    fallback hostname                        : localhost
    default compression method               : zstd
    default DNSSEC mode                      : no
    default DNS-over-TLS mode                : no
    default mDNS mode                        : yes
    default LLMNR mode                       : yes
    default DNS servers                      :
    default NTP servers                      :
    default net.naming_scheme= value         : latest
    default KillUserProcesses= value         : True
    default locale                           : C.UTF-8
    default nspawn locale                    : C.UTF-8
    default status unit format               : name
    default user $PATH                       : (same as system services)
    systemd service watchdog                 : 3min
    time epoch                               : 946681200 (1999-12-31T23:00:00+00:00)

  Features
    enabled                                  : SECCOMP, blkid, libfdisk, pcre2, zstd, coredump, hwdb, initrd, kernel-install, logind, machined, networkd, nss-myhostname, nss-mymachines, oomd, portabled, pstore, repart, sysext, systemd-analyze, storagetm, timesyncd, userdb, vconsole, vmspawn, xdg-autostart, dmi, nscd, kmod, adm group, wheel group, gshadow, link-udev-shared, link-systemctl-shared, link-networkd-shared, link-timesyncd-shared, link-journalctl-shared, link-boot-shared, link-portabled-shared
    disabled                                 : ACL, AUDIT, AppArmor, IMA, PAM, SELinux, SMACK, elfutils, gcrypt, gnutls, libbpf, libcryptsetup, libcryptsetup-plugins, libcurl, libfido2, libidn, libidn2, libiptc, microhttpd, openssl, p11kit, passwdqc, pwquality, qrencode, tpm2, xkbcommon, lz4, xz, zlib, bzip2, backlight, binfmt, bootloader, bpf-framework, efi, environment.d, firstboot, hibernate, homed, hostnamed, importd, localed, nss-resolve, nss-systemd, quotacheck, randomseed, resolve, rfkill, sysupdate, sysusers, timedated, tmpfiles, idn, polkit, legacy-pkla, xenctrl, dbus, glib, tpm, man pages, html pages, man page indices, SysV compat, compat-mutable-uid-boundaries, utmp, ldconfig, debug hashmap, debug mmap cache, debug siphash, trace logging, slow tests, fuzz tests, install tests, first-boot-full-preset, fexecve, standalone-binaries, coverage, static-libsystemd, static-libudev, cryptolib, DNS-over-TLS

  User defined options
    buildtype                                : debug
    libdir                                   : /lib
    prefix                                   : //
    acl                                      : false
    apparmor                                 : false
    audit                                    : false
    backlight                                : false
    bashcompletiondir                        : /etc/bash_completion.d
    binfmt                                   : false
    blkid                                    : true
    bzip2                                    : false
    certificate-root                         : /etc/ssl
    clock-valid-range-usec-max               : 3153600000000000
    coredump                                 : true
    dbus                                     : false
    dbuspolicydir                            : /etc/dbus-1/system.d
    dbussessionservicedir                    : /usr/share/dbus-1/services
    dbussystemservicedir                     : /usr/share/dbus-1/system-services
    default-dnssec                           : no
    dns-servers                              :
    efi                                      : false
    elfutils                                 : false
    environment-d                            : false
    fallback-hostname                        : localhost
    firstboot                                : false
    gcrypt                                   : false
    glib                                     : false
    gnutls                                   : false
    hibernate                                : false
    hostnamed                                : false
    html                                     : false
    hwdb                                     : true
    idn                                      : false
    ima                                      : false
    importd                                  : false
    kmod                                     : true
    ldconfig                                 : false
    libcryptsetup                            : false
    libcurl                                  : false
    libidn                                   : false
    libiptc                                  : false
    localed                                  : false
    logind                                   : true
    lz4                                      : false
    machined                                 : true
    man                                      : false
    microhttpd                               : false
    networkd                                 : true
    nss-myhostname                           : true
    nss-systemd                              : false
    ntp-servers                              :
    openssl                                  : false
    pam                                      : false
    pcre2                                    : true
    polkit                                   : false
    qrencode                                 : false
    quotacheck                               : false
    randomseed                               : false
    rc-local                                 :
    remote                                   : false
    resolve                                  : false
    rfkill                                   : false
    seccomp                                  : true
    selinux                                  : false
    smack                                    : false
    split-bin                                : true
    split-usr                                : true
    sysusers                                 : false
    sysvinit-path                            :
    sysvrcnd-path                            :
    time-epoch                               : 946681200
    timedated                                : false
    timesyncd                                : true
    tmpfiles                                 : false
    tpm                                      : false
    tpm2                                     : false
    utmp                                     : false
    vconsole                                 : true
    xkbcommon                                : false
    xz                                       : false
    zlib                                     : false
    zstd                                     : true
```
    
The problem can be circumvented by building with `-D mountfsd=false`

**Comment by myrkr** [comment]
With 6a68270c9795fd26b897a7cc33a8b39ae72d6d2c applied the meson run fails in this cause because the mountfsd refers to an unbound variable:

```
../src/mountfsd/meson.build:12:0: ERROR: Unknown variable "common_libs".
```

at least here the problem is better shown.

**Comment by YHNdnzj** [comment]
> The problem is likely caused by src/importd providing a variable common_libs:
> ...
> and src/mountfsd using that lib

Yeah, this seems wrong. I don't see such dependency anywhere.

**Comment by myrkr** [comment]
I checked now with the main branch instead of v256-rc1. I can no longer reproduce the problem, likely fixed by fea14ca15eb4b3873f85dbb5e24dc2929448b143
